staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -476,6 +559,51 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
 	return curObj, oldObj, nil
 }
 
+// isInitialEventsEndBookmarkSent same as isInitialEventsEndBookmarkSentLocked
+// just acquires the lock.
+func (wc *watchChan) isInitialEventsEndBookmarkSent() bool {
+	wc.initialEventsEndBookmarkSentMutex.RLock()
+	defer wc.initialEventsEndBookmarkSentMutex.RUnlock()
+	return wc.isInitialEventsEndBookmarkSentLocked()
+}
+
+// isInitialEventsEndBookmarkSentLocked checks if
+// the given watchChan has sent an annotated
+// bookmark event marking the end of streaming.
+//
+// For more information see the watch-list feature.
+func (wc *watchChan) isInitialEventsEndBookmarkSentLocked() bool {
+	return wc.initialEventsEndBookmarkSent
+}
+
+// setInitialEventsEndBookmarkSent marks that the given watchChan
+// has sent an annotated bookmark event.
+func (wc *watchChan) setInitialEventsEndBookmarkSent(event *watch.Event) {
+	if event.Type == watch.Bookmark {
+		// minimize contention and allow concurrent read access
+		// by first checking if there is work to do.
+		wc.initialEventsEndBookmarkSentMutex.RLock()
+		initialEventsEndBookmarkSent := wc.isInitialEventsEndBookmarkSentLocked()
+		wc.initialEventsEndBookmarkSentMutex.RUnlock()
+		if initialEventsEndBookmarkSent {
+			return
+		}
+		// ok, now we now that a write operation is necessary
+		// also note that a write lock is rare
+		// compared to the frequency of read access
+		// in fact this code we be executed
+		// exactly once
+		wc.initialEventsEndBookmarkSentMutex.Lock()
+		defer wc.initialEventsEndBookmarkSentMutex.Unlock()
+		wc.initialEventsEndBookmarkSent = true
+	}
+}
+
+func (wc *watchChan) getCurrentResourceVersionFromStorage() (uint64, error) {
```

Comment:
We alredy have that implemented for cacher - we should refactor it and share the code - it should be exactly the same.

Ideally in a separate preparation PR - might be the same as the one I asked for above]

Reply:
yes, I still need to figure out how to pass the `newListFunc` or something similar.